File: packages/sdks/server-sdk/src/client/ApiClient.ts
/**
 * HTTP API Client using Axios
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse, AxiosError } from 'axios';
import { ulid } from 'ulid';
import { Logger } from '../utils/logger';
import { ErrorCode, createError, isGatrixSDKError } from '../utils/errors';
import { ApiResponse } from '../types/api';
import { RetryConfig } from '../types/config';
import { SdkMetrics } from '../utils/sdkMetrics';
import { sleep } from '../utils/time';

export interface ApiClientConfig {
  baseURL: string;
  apiToken: string;
  applicationName: string;
  environment?: string; // Default environment for single-environment mode
  timeout?: number;
  logger?: Logger;
  retry?: RetryConfig;
  metrics?: SdkMetrics; // Optional SDK metrics collector
}

const DEFAULT_RETRY_CONFIG: Required<RetryConfig> = {
  enabled: true,
  maxRetries: 10,
  retryDelay: 2000, // Initial delay: 2 seconds
  retryDelayMultiplier: 2, // Exponential backoff: 2s -> 4s -> 8s -> 10s (max)
  maxRetryDelay: 10000, // Max delay: 10 seconds
  retryableStatusCodes: [408, 429, 500, 502, 503, 504],
};

export class ApiClient {
  private client: AxiosInstance;
  private logger: Logger;
  private retryConfig: Required<RetryConfig>;
  private metrics?: SdkMetrics;
  private etagStore: Map<string, string>;
  private bodyCache: Map<string, ApiResponse<any>>;

  // Connection recovery callback system
  private connectionRecoveryCallbacks: Array<() => void> = [];
  private lastRecoveryTriggerTime: number = 0;
  private recoveryThrottleMs: number = 5000; // Throttle: max once per 5 seconds

  constructor(config: ApiClientConfig) {
    this.logger = config.logger || new Logger();
    this.retryConfig = { ...DEFAULT_RETRY_CONFIG, ...config.retry };
    this.metrics = config.metrics;
    this.etagStore = new Map();
    this.bodyCache = new Map();

    // Build headers with optional environment
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'X-API-Token': config.apiToken,
      'X-Application-Name': config.applicationName,
    };

    // Add environment header for single-environment mode
    if (config.environment) {
      headers['X-Environment'] = config.environment;
    }

    // Create axios instance
    this.client = axios.create({
      baseURL: config.baseURL,
      timeout: config.timeout || 30000,
      headers,
      // Treat 304 Not Modified as a successful response so we can handle ETag logic
      validateStatus: (status) => (status >= 200 && status < 300) || status === 304,
    });

    // Request interceptor
    this.client.interceptors.request.use(
      (requestConfig) => {
        // Record start time for latency metrics
        (requestConfig as any).__sdkStartTime = process.hrtime.bigint();

        this.logger.debug('API Request', {
          method: requestConfig.method?.toUpperCase(),
          url: requestConfig.url,
          params: requestConfig.params,
        });
        return requestConfig;
      },
      (error) => {
        this.logger.error('API Request Error', { error: error.message });
        return Promise.reject(error);
      }
    );

    // Response interceptor
    this.client.interceptors.response.use(
      (response) => {
        this.logger.debug('API Response', {
          status: response.status,
          url: response.config.url,
        });

        // Observe HTTP duration and increment counters
        try {
          const started = (response.config as any).__sdkStartTime as bigint | undefined;
          if (started && this.metrics) {
            const elapsed = Number(process.hrtime.bigint() - started) / 1e9;
            const method = response.config.method?.toUpperCase() || 'GET';
            const route = response.config.url || 'unknown';
            // Use SDK metrics generic error counter and custom HTTP helpers if present
            this.metrics.observeHttpDuration?.(method, route, response.status, elapsed);
            this.metrics.incHttpRequestsTotal?.(method, route, response.status);
          }
        } catch {
          // ignore metrics failures
        }

        return response;
      },
      (error: AxiosError) => {
        try {
          const cfg = error.config as any;
          const started = cfg?.__sdkStartTime as bigint | undefined;
          if (started && this.metrics) {
            const elapsed = Number(process.hrtime.bigint() - started) / 1e9;
            const method = (cfg?.method?.toUpperCase?.() as string) || 'GET';
            const route = (cfg?.url as string) || 'unknown';
            const status = error.response?.status || 0;
            this.metrics.observeHttpDuration?.(method, route, status, elapsed);
            this.metrics.incHttpRequestsTotal?.(method, route, status);
            this.metrics.incError('http', 'request');
          }
        } catch {
          // ignore metrics failures
        }

        // Don't call handleError here - it will be called in executeWithRetry on final failure
        // This prevents duplicate error logs during retry cycles
        return Promise.reject(error);
      }
    );
  }

  /**
   * Handle API errors
   */
  private handleError(error: AxiosError): void {
    if (error.response) {
      // Server responded with error status
      const status = error.response.status;
      const data = error.response.data as any;

      this.logger.error('API Error Response', {
        status,
        url: error.config?.url,
        error: data?.error || data?.message,
      });

      if (status === 401) {
        throw createError(
          ErrorCode.AUTH_FAILED,
          'Authentication failed. Please check your API token.',
          status,
          data
        );
      } else if (status >= 400 && status < 500) {
        throw createError(
          ErrorCode.INVALID_PARAMETERS,
          data?.error?.message || data?.message || 'Invalid request parameters',
          status,
          data
        );
      } else if (status >= 500) {
        throw createError(
          ErrorCode.API_ERROR,
          data?.error?.message || data?.message || 'Server error occurred',
          status,
          data
        );
      }
    } else if (error.request) {
      // Request was made but no response received
      this.logger.error('Network Error', {
        url: error.config?.url,
        message: error.message,
      });

      throw createError(
        ErrorCode.NETWORK_ERROR,
        'Network error. Please check your connection.',
        undefined,
        { message: error.message }
      );
    } else {
      // Something else happened
      this.logger.error('Request Setup Error', { message: error.message });

      throw createError(
        ErrorCode.API_ERROR,
        error.message || 'An unexpected error occurred',
        undefined,
        { message: error.message }
      );
    }
  }

  /**
   * Check if error is retryable
   */
  private isRetryableError(error: any): boolean {
    if (!this.retryConfig.enabled) {
      return false;
    }

    // Handle GatrixSDKError (as errors are wrapped in interceptor)
    if (isGatrixSDKError(error)) {
      if (error.statusCode) {
        return this.retryConfig.retryableStatusCodes.includes(error.statusCode);
      }
      // Network errors are retryable
      return error.code === ErrorCode.NETWORK_ERROR;
    }

    // Handle raw AxiosError
    const axiosError = error as AxiosError;

    // Network errors are retryable
    if (!axiosError.response) {
      return true;
    }

    // Check if status code is retryable
    const status = axiosError.response.status;
    return this.retryConfig.retryableStatusCodes.includes(status);
  }

  /**
   * Calculate retry delay with exponential backoff
   */
  private calculateRetryDelay(attemptNumber: number): number {
    const delay =
      this.retryConfig.retryDelay * Math.pow(this.retryConfig.retryDelayMultiplier, attemptNumber);
    return Math.min(delay, this.retryConfig.maxRetryDelay);
  }

  /**
   * Build a stable cache key for ETag storage based on URL and query params
   */
  private buildCacheKey(url: string, config?: AxiosRequestConfig): string {
    if (!config || !config.params) {
      return url;
    }

    try {
      const params = config.params as Record<string, any>;
      const sortedKeys = Object.keys(params).sort();
      const normalized: Record<string, any> = {};

      for (const key of sortedKeys) {
        const value = params[key];
        if (value === undefined) {
          continue;
        }
        normalized[key] = value;
      }

      return `${url}?${JSON.stringify(normalized)}`;
    } catch {
      // Fallback to URL only if params cannot be serialized safely
      return url;
    }
  }

  /**
   * Execute request with retry logic
   * Supports infinite retries when maxRetries is -1
   */
  private async executeWithRetry<T>(
    fn: () => Promise<AxiosResponse<ApiResponse<T>>>,
    context: { method: string; url: string }
  ): Promise<ApiResponse<T>> {
    let lastError: any;
    const isInfiniteRetry = this.retryConfig.maxRetries === -1;
    const maxAttempts = isInfiniteRetry ? Number.MAX_SAFE_INTEGER : this.retryConfig.maxRetries;

    for (let attempt = 0; attempt <= maxAttempts; attempt++) {
      try {
        const response = await fn();

        // Log recovery if this was a retry attempt (attempt > 0 means we had at least one failure)
        if (attempt > 0) {
          this.logger.info('*** API connection restored ***', {
            method: context.method,
            url: context.url,
            retriesNeeded: attempt,
          });

          // Trigger connection recovery callbacks (throttled)
          this.triggerConnectionRecoveryCallbacks();
        }

        return response.data;
      } catch (error) {
        lastError = error;

        // Check if this is the final failure (not retryable or last attempt)
        const isFinalFailure = !this.isRetryableError(lastError) || attempt === maxAttempts;

        if (isFinalFailure) {
          // Log detailed error only on final failure
          this.logger.error('Request failed permanently', {
            method: context.method,
            url: context.url,
            attempt: attempt + 1,
            maxRetries: isInfiniteRetry ? 'infinite' : this.retryConfig.maxRetries,
            error: this.extractDetailedErrorMessage(lastError),
          });

          // Call handleError to transform the error before throwing
          this.handleError(lastError as AxiosError);
          throw error;
        }

        // Calculate delay and retry
        const delay = this.calculateRetryDelay(attempt);
        this.logger.warn('Request failed, retrying...', {
          method: context.method,
          url: context.url,
          attempt: attempt + 1,
          maxRetries: isInfiniteRetry ? 'infinite' : this.retryConfig.maxRetries,
          retryDelay: delay,
          error: this.extractDetailedErrorMessage(lastError),
        });

        await sleep(delay);
      }
    }

    // This should never happen, but TypeScript needs it
    throw lastError;
  }

  /**
   * Extract a detailed error message from various error types
   */
  private extractDetailedErrorMessage(error: any): string {
    // Handle AxiosError with response data
    if (error.response?.data) {
      const data = error.response.data;
      if (data.error?.message) {
        return data.error.message;
      }
      if (data.message) {
        return data.message;
      }
    }

    // Handle network errors with more context
    if (error.code) {
      // Axios error codes like ECONNREFUSED, ENOTFOUND, ETIMEDOUT, etc.
      return `${error.code}: ${error.message || 'Unknown error'}`;
    }

    // Default to error message
    return error.message || 'Unknown error';
  }

  /**
   * GET request (with retry)
   */
  async get<T = any>(url: string, config?: AxiosRequestConfig): Promise<ApiResponse<T>> {
    const requestConfig: AxiosRequestConfig = { ...(config || {}) };
    const headers: Record<string, any> = { ...(requestConfig.headers || {}) };

    const cacheKey = this.buildCacheKey(url, requestConfig);
    const cachedEtag = this.etagStore.get(cacheKey);
    if (cachedEtag) {
      headers['If-None-Match'] = cachedEtag;
    }

    requestConfig.headers = headers;

    return this.executeWithRetry(
      async () => {
        const response = await this.client.get<ApiResponse<T>>(url, requestConfig);

        // Handle 304 Not Modified using cached response body
        if (response.status === 304) {
          const cachedBody = this.bodyCache.get(cacheKey) as ApiResponse<T> | undefined;
          if (cachedBody) {
            return {
              ...response,
              status: 200,
              data: cachedBody,
            } as AxiosResponse<ApiResponse<T>>;
          }

          // No cached body but got 304: fall back to a fresh request without conditional header
          const retryConfig: AxiosRequestConfig = {
            ...requestConfig,
            headers: { ...headers },
          };
          delete (retryConfig.headers as any)['If-None-Match'];
          delete (retryConfig.headers as any)['if-none-match'];

          const freshResponse = await this.client.get<ApiResponse<T>>(url, retryConfig);

          const freshEtag = (freshResponse.headers as any)?.etag as string | undefined;
          if (freshEtag) {
            this.etagStore.set(cacheKey, freshEtag);
            this.bodyCache.set(cacheKey, freshResponse.data);
          }

          return freshResponse;
        }

        // Normal 2xx response: capture ETag and cache body if present
        const etag = (response.headers as any)?.etag as string | undefined;
        if (etag) {
          this.etagStore.set(cacheKey, etag);
          this.bodyCache.set(cacheKey, response.data);
        }

        return response;
      },
      { method: 'GET', url }
    );
  }

  /**
   * POST request (with retry and auto-generated requestId in header)
   */
  async post<T = any>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ): Promise<ApiResponse<T>> {
    // Auto-generate requestId if not provided in headers
    const requestConfig = config || {};
    const headers = requestConfig.headers || {};

    if (!headers['x-request-id']) {
      headers['x-request-id'] = ulid();
      this.logger.debug('Auto-generated requestId', {
        requestId: headers['x-request-id'],
        url,
      });
    }

    requestConfig.headers = headers;

    return this.executeWithRetry(() => this.client.post<ApiResponse<T>>(url, data, requestConfig), {
      method: 'POST',
      url,
    });
  }

  /**
   * POST request without retry (for operations like unregister that should not retry)
   */
  async postNoRetry<T = any>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ): Promise<ApiResponse<T>> {
    // Auto-generate requestId if not provided in headers
    const requestConfig = config || {};
    const headers = requestConfig.headers || {};

    if (!headers['x-request-id']) {
      headers['x-request-id'] = ulid();
      this.logger.debug('Auto-generated requestId', {
        requestId: headers['x-request-id'],
        url,
      });
    }

    requestConfig.headers = headers;

    try {
      const response = await this.client.post<ApiResponse<T>>(url, data, requestConfig);
      return response.data;
    } catch (error) {
      this.handleError(error as AxiosError);
      throw error;
    }
  }

  /**
   * PUT request (with retry and auto-generated requestId in header)
   */
  async put<T = any>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ): Promise<ApiResponse<T>> {
    // Auto-generate requestId if not provided in headers
    const requestConfig = config || {};
    const headers = requestConfig.headers || {};

    if (!headers['x-request-id']) {
      headers['x-request-id'] = ulid();
      this.logger.debug('Auto-generated requestId', {
        requestId: headers['x-request-id'],
        url,
      });
    }

    requestConfig.headers = headers;

    return this.executeWithRetry(() => this.client.put<ApiResponse<T>>(url, data, requestConfig), {
      method: 'PUT',
      url,
    });
  }

  /**
   * PATCH request (with retry and auto-generated requestId in header)
   */
  async patch<T = any>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ): Promise<ApiResponse<T>> {
    // Auto-generate requestId if not provided in headers
    const requestConfig = config || {};
    const headers = requestConfig.headers || {};

    if (!headers['x-request-id']) {
      headers['x-request-id'] = ulid();
      this.logger.debug('Auto-generated requestId', {
        requestId: headers['x-request-id'],
        url,
      });
    }

    requestConfig.headers = headers;

    return this.executeWithRetry(
      () => this.client.patch<ApiResponse<T>>(url, data, requestConfig),
      { method: 'PATCH', url }
    );
  }

  /**
   * DELETE request (with retry)
   */
  async delete<T = any>(url: string, config?: AxiosRequestConfig): Promise<ApiResponse<T>> {
    return this.executeWithRetry(() => this.client.delete<ApiResponse<T>>(url, config), {
      method: 'DELETE',
      url,
    });
  }

  /**
   * Get axios instance (for advanced usage)
   */
  getAxiosInstance(): AxiosInstance {
    return this.client;
  }

  /**
   * Invalidate ETag cache for a specific URL pattern
   * This forces the next request to fetch fresh data instead of returning cached 304 response
   * @param urlPattern URL or URL pattern (partial match) to invalidate
   */
  invalidateEtagCache(urlPattern: string): void {
    const keysToDelete: string[] = [];
    for (const key of this.etagStore.keys()) {
      if (key.includes(urlPattern)) {
        keysToDelete.push(key);
      }
    }
    for (const key of keysToDelete) {
      this.etagStore.delete(key);
      this.bodyCache.delete(key);
    }
    if (keysToDelete.length > 0) {
      this.logger.debug('Invalidated ETag cache', {
        urlPattern,
        count: keysToDelete.length,
      });
    }
  }

  /**
   * Clear all ETag caches
   */
  clearAllEtagCache(): void {
    const count = this.etagStore.size;
    this.etagStore.clear();
    this.bodyCache.clear();
    this.logger.debug('Cleared all ETag cache', { count });
  }

  /**
   * Register a callback to be called when connection is recovered after retries
   * Callbacks are throttled to prevent multiple triggers in quick succession
   * @param callback Function to call on connection recovery
   * @returns Function to unregister the callback
   */
  onConnectionRecovered(callback: () => void): () => void {
    this.connectionRecoveryCallbacks.push(callback);
    return () => {
      const index = this.connectionRecoveryCallbacks.indexOf(callback);
      if (index !== -1) {
        this.connectionRecoveryCallbacks.splice(index, 1);
      }
    };
  }

  /**
   * Trigger connection recovery callbacks (with throttle)
   * Called internally when a request succeeds after retries
   */
  private triggerConnectionRecoveryCallbacks(): void {
    const now = Date.now();

    // Throttle: only trigger if enough time has passed since last trigger
    if (now - this.lastRecoveryTriggerTime < this.recoveryThrottleMs) {
      this.logger.debug('Connection recovery callback throttled', {
        timeSinceLastTrigger: now - this.lastRecoveryTriggerTime,
        throttleMs: this.recoveryThrottleMs,
      });
      return;
    }

    this.lastRecoveryTriggerTime = now;

    // Call all registered callbacks asynchronously (don't block the request)
    if (this.connectionRecoveryCallbacks.length > 0) {
      this.logger.debug('Triggering connection recovery callbacks', {
        count: this.connectionRecoveryCallbacks.length,
      });

      // Execute callbacks in a microtask to avoid blocking
      Promise.resolve().then(() => {
        for (const callback of this.connectionRecoveryCallbacks) {
          try {
            callback();
          } catch (error: any) {
            this.logger.error('Error in connection recovery callback', {
              error: error.message,
            });
          }
        }
      });
    }
  }
}

File: packages/sdks/server-sdks/gatrix-go-server-sdk/client/api_client.go
package client

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"sync"
	"time"

	"github.com/gatrix/gatrix-go-server-sdk/types"
)

// RetryConfig holds HTTP retry settings
type RetryConfig struct {
	Enabled              *bool
	MaxRetries           int
	RetryDelay           int
	RetryDelayMultiplier int
	MaxRetryDelay        int
	RetryableStatusCodes []int
}

// ApiClientConfig holds configuration for creating an ApiClient
type ApiClientConfig struct {
	BaseURL         string
	APIToken        string
	ApplicationName string
	Timeout         time.Duration
	Logger          types.Logger
	Retry           *RetryConfig
}

var defaultRetryValues = RetryConfig{
	Enabled:              boolP(true),
	MaxRetries:           10,
	RetryDelay:           2000,
	RetryDelayMultiplier: 2,
	MaxRetryDelay:        10000,
	RetryableStatusCodes: []int{408, 429, 500, 502, 503, 504},
}

func boolP(v bool) *bool { return &v }

// ApiClient is the HTTP client for Gatrix API
type ApiClient struct {
	httpClient      *http.Client
	baseURL         string
	apiToken        string
	applicationName string
	logger          types.Logger
	retryConfig     RetryConfig
	etagStore       map[string]string
	bodyCache       map[string]json.RawMessage
	mu              sync.RWMutex
}

// ApiResponse is the standard API response
type ApiResponse struct {
	Success bool            `json:"success"`
	Data    json.RawMessage `json:"data,omitempty"`
	Error   *ApiError       `json:"error,omitempty"`
}

// ApiError represents an API error
type ApiError struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// NewApiClient creates a new API client
func NewApiClient(config ApiClientConfig) *ApiClient {
	timeout := config.Timeout
	if timeout == 0 {
		timeout = 30 * time.Second
	}
	logger := config.Logger
	if logger == nil {
		logger = types.NewDefaultLogger(types.LogLevelInfo)
	}

	retry := defaultRetryValues
	if config.Retry != nil {
		if config.Retry.Enabled != nil {
			retry.Enabled = config.Retry.Enabled
		}
		if config.Retry.MaxRetries > 0 {
			retry.MaxRetries = config.Retry.MaxRetries
		}
		if config.Retry.RetryDelay > 0 {
			retry.RetryDelay = config.Retry.RetryDelay
		}
		if config.Retry.RetryDelayMultiplier > 0 {
			retry.RetryDelayMultiplier = config.Retry.RetryDelayMultiplier
		}
		if config.Retry.MaxRetryDelay > 0 {
			retry.MaxRetryDelay = config.Retry.MaxRetryDelay
		}
		if len(config.Retry.RetryableStatusCodes) > 0 {
			retry.RetryableStatusCodes = config.Retry.RetryableStatusCodes
		}
	}

	return &ApiClient{
		httpClient:      &http.Client{Timeout: timeout},
		baseURL:         config.BaseURL,
		apiToken:        config.APIToken,
		applicationName: config.ApplicationName,
		logger:          logger,
		retryConfig:     retry,
		etagStore:       make(map[string]string),
		bodyCache:       make(map[string]json.RawMessage),
	}
}

// Get performs a GET request with ETag caching and retry
func (c *ApiClient) Get(path string) (*ApiResponse, error) {
	return c.doRequest("GET", path, nil)
}

// Post performs a POST request with retry
func (c *ApiClient) Post(path string, body interface{}) (*ApiResponse, error) {
	return c.doRequest("POST", path, body)
}

// Put performs a PUT request with retry
func (c *ApiClient) Put(path string, body interface{}) (*ApiResponse, error) {
	return c.doRequest("PUT", path, body)
}

// Delete performs a DELETE request with retry
func (c *ApiClient) Delete(path string) (*ApiResponse, error) {
	return c.doRequest("DELETE", path, nil)
}

// PostNoRetry performs a POST without retry (for shutdown operations)
func (c *ApiClient) PostNoRetry(path string, body interface{}) (*ApiResponse, error) {
	return c.doSingleRequest("POST", path, body)
}

// InvalidateEtagCache removes cached ETag for a URL pattern
func (c *ApiClient) InvalidateEtagCache(urlPattern string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	for key := range c.etagStore {
		if len(key) >= len(urlPattern) && key[:len(urlPattern)] == urlPattern {
			delete(c.etagStore, key)
			delete(c.bodyCache, key)
		}
	}
}

func (c *ApiClient) doRequest(method, path string, body interface{}) (*ApiResponse, error) {
	enabled := true
	if c.retryConfig.Enabled != nil {
		enabled = *c.retryConfig.Enabled
	}
	maxAttempts := c.retryConfig.MaxRetries
	if !enabled {
		maxAttempts = 0
	}

	var lastErr error
	for attempt := 0; attempt <= maxAttempts; attempt++ {
		resp, err := c.doSingleRequest(method, path, body)
		if err == nil {
			return resp, nil
		}
		lastErr = err
		if !c.isRetryableError(err) {
			return nil, err
		}
		if attempt < maxAttempts {
			delay := c.calculateRetryDelay(attempt)
			c.logger.Warn("Request failed, retrying...", map[string]interface{}{
				"method": method, "path": path, "attempt": attempt + 1,
				"retryDelay": delay, "error": err.Error(),
			})
			time.Sleep(time.Duration(delay) * time.Millisecond)
		}
	}
	return nil, lastErr
}

func (c *ApiClient) doSingleRequest(method, path string, body interface{}) (*ApiResponse, error) {
	url := c.baseURL + path
	var bodyReader io.Reader
	if body != nil {
		bodyBytes, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		bodyReader = bytes.NewReader(bodyBytes)
	}

	req, err := http.NewRequest(method, url, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-API-Token", c.apiToken)
	req.Header.Set("X-Application-Name", c.applicationName)
	req.Header.Set("X-SDK-Version", fmt.Sprintf("%s/%s", types.SDKName, types.SDKVersion))

	if method == "GET" {
		c.mu.RLock()
		etag, hasEtag := c.etagStore[path]
		c.mu.RUnlock()
		if hasEtag {
			req.Header.Set("If-None-Match", etag)
		}
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, &types.GatrixError{Code: types.ErrorCodeNetworkError, Message: fmt.Sprintf("Network error: %s", err.Error())}
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotModified {
		c.mu.RLock()
		cachedBody, hasCached := c.bodyCache[path]
		c.mu.RUnlock()
		if hasCached {
			var apiResp ApiResponse
			if err := json.Unmarshal(cachedBody, &apiResp); err == nil {
				return &apiResp, nil
			}
		}
	}

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	if resp.StatusCode >= 400 {
		var apiResp ApiResponse
		_ = json.Unmarshal(respBody, &apiResp)
		errMsg := "API error"
		errCode := types.ErrorCodeAPIError
		if apiResp.Error != nil {
			errMsg = apiResp.Error.Message
		}
		if resp.StatusCode == 401 {
			errCode = types.ErrorCodeAuthFailed
			errMsg = "Authentication failed. Please check your API token."
		} else if resp.StatusCode < 500 {
			errCode = types.ErrorCodeInvalidParams
		}
		return nil, &types.GatrixError{Code: errCode, Message: errMsg, StatusCode: resp.StatusCode}
	}

	if method == "GET" {
		etag := resp.Header.Get("ETag")
		if etag != "" {
			c.mu.Lock()
			c.etagStore[path] = etag
			c.bodyCache[path] = json.RawMessage(respBody)
			c.mu.Unlock()
		}
	}

	var apiResp ApiResponse
	if err := json.Unmarshal(respBody, &apiResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}
	return &apiResp, nil
}

func (c *ApiClient) isRetryableError(err error) bool {
	if gatrixErr, ok := err.(*types.GatrixError); ok {
		if gatrixErr.Code == types.ErrorCodeNetworkError {
			return true
		}
		if gatrixErr.StatusCode > 0 {
			for _, code := range c.retryConfig.RetryableStatusCodes {
				if gatrixErr.StatusCode == code {
					return true
				}
			}
		}
		return false
	}
	return true
}

func (c *ApiClient) calculateRetryDelay(attempt int) int {
	delay := float64(c.retryConfig.RetryDelay) * math.Pow(float64(c.retryConfig.RetryDelayMultiplier), float64(attempt))
	if delay > float64(c.retryConfig.MaxRetryDelay) {
		return c.retryConfig.MaxRetryDelay
	}
	return int(delay)
}
